### PR TITLE
Script to dump content into an S3 bucket

### PR DESCRIPTION
**Pull request by mmulich**
See the readme for basic information. We can adjust it as needed.

### CODE REVIEW

File: dump/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+1. Elementary Algebra (col12116) - 0889907c-f0ef-496a-bcb8-2a5bb121717f@8.42
+1. Intermediate Algebra (col12119) - 02776133-d49d-49cb-bfaa-67c7f61b25a1@8.13
+1. Prealgebra (col11756) - caa57dab-41c7-455e-bd6f-f443cda5519c@19.5
+
+## Installation
+
+```sh
+pip install -r requirements.txt
+```
+
```
Comment (karenc) [inline dump/README.md:16]: May be we can add a link to boto3 quick start for the aws set up stuff https://github.com/boto/boto3#quick-start

File: dump/dump-to-bucket.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+    temperature = 'raw'
+    format_ = 'json'
+    url = f'{base_raw_url}.{format_}{raw_postfix}'
+    # debug(f'Requesting {temperature} JSON {T.bold}{type_}{T.normal} at {T.yellow}{url}{T.normal}')
+    # resp = session.get(url)
+    # yield io.BytesIO(resp.content), 'application/json', f'{temperature}-{type_}-json', [ident_hash_seq[-1]]
+    yield get_content(url, type_, format_, temperature, ident_hash_seq, 'application/json', -1)
+
+    # Save the raw json for later
+    raw_json = resp.json()
```
Comment (karenc) [inline dump/dump-to-bucket.py:206]: Mike got the same error:

```
Dumping raw-book-json into the bucket at "raw/02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10.json" (application/json)
Traceback (most recent call last):
  File "./dump-to-bucket.py", line 296, in <module>
    main()
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "./dump-to-bucket.py", line 292, in main
    dump_in_bucket(scrape(book, host), (bucket, region))
  File "./dump-to-bucket.py", line 268, in dump_in_bucket
    for item in items:
  File "./dump-to-bucket.py", line 122, in scrape
    for item in scrape_version(id, version, host, visited_locs):
  File "./dump-to-bucket.py", line 216, in scrape_version
    raw_json = resp.json()
UnboundLocalError: local variable 'resp' referenced before assignment
```

At first I thought I can add `resp = yield get_content(...` on line 213 but that's not it...  Need to read more

File: dump/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+blessings
+boto3
+click
+requests
```
Comment (karenc) [inline dump/requirements.txt:5]: Missing `cnx-common` in this list

**Review by karenc — CHANGES_REQUESTED** [review]

File: dump/dump-to-bucket.py (hunk 2 above)
Comment (karenc) [reply dump/dump-to-bucket.py:206]: Ok I get it!  The `resp.json()` lines just need to be removed or commented out.

File: dump/dump-to-bucket.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+def gen_filepath(type_, ident):
+    """Given a content type and an ids structure
+    produce the S3 filepath to the object.
+
+    For json and html the ``ident`` sequence of one or more tuples of UUID and version.
+    For a resource the ``ident`` is the SHA1.
+
+    """
+    if not type_.startswith('resource'):
+        ids = [join_ident_hash(*i) for i in ident]
```
Comment (karenc) [inline dump/dump-to-bucket.py:62]:
```
Traceback (most recent call last):
  File "./dump-to-bucket.py", line 290, in <module>
    main()
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "./dump-to-bucket.py", line 286, in main
    dump_in_bucket(scrape(book, host), (bucket, region))
  File "./dump-to-bucket.py", line 264, in dump_in_bucket
    key = gen_filepath(type, ident)
  File "./dump-to-bucket.py", line 62, in gen_filepath
    ids = [join_ident_hash(*i) for i in ident]
  File "./dump-to-bucket.py", line 62, in <listcomp>
    ids = [join_ident_hash(*i) for i in ident]
TypeError: join_ident_hash() missing 1 required positional argument: 'version'
```

I did a bit of debugging and `i` looks like this:

```
latest version of requested book: 02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10
Requesting raw json book at https://archive.cnx.org/contents/02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10.json?as_collated=0
i: ('02776133-d49d-49cb-bfaa-67c7f61b25a1', '8.10')
Dumping raw-book-json into the bucket at "raw/02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10.json" (application/json)
Requesting baked json book at https://archive.cnx.org/contents/02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10.json
i: ('02776133-d49d-49cb-bfaa-67c7f61b25a1', '8.10')
Dumping baked-book-json into the bucket at "baked/02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10.json" (application/json)
Requesting raw html book at https://archive.cnx.org/contents/02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10.html?as_collated=0
i: ('02776133-d49d-49cb-bfaa-67c7f61b25a1', '8.10')
Dumping raw-book-json into the bucket at "raw/02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10.json" (text/html)
Requesting baked html book at https://archive.cnx.org/contents/02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10.html
i: [('02776133-d49d-49cb-bfaa-67c7f61b25a1', '8.10')]
```

File: dump/dump-to-bucket.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    yield get_content(url, type_, format_, temperature, ident_hash_seq, 'text/html', -1)
+
+    # Request the BAKED HTML
+    temperature = 'baked'
+    format_ = 'html'
+    url = f'{base_baked_url}.html'
+    # debug(f'Requesting {temperature} HTML {T.bold}{type_}{T.normal} at {T.yellow}{url}{T.normal}')
+    # resp = session.get(url)
+    # yield io.BytesIO(resp.content), 'text/html', f'{temperature}-{type_}-html', ident_hash_seq
+    yield get_content(url, type_, format_, temperature, ident_hash_seq, 'text/html')
```
Comment (karenc) [inline dump/dump-to-bucket.py]: This is the line that caused the error:

```
Traceback (most recent call last):
  File "./dump-to-bucket.py", line 290, in <module>
    main()
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "./dump-to-bucket.py", line 286, in main
    dump_in_bucket(scrape(book, host), (bucket, region))
  File "./dump-to-bucket.py", line 264, in dump_in_bucket
    key = gen_filepath(type, ident)
  File "./dump-to-bucket.py", line 62, in gen_filepath
    ids = [join_ident_hash(*i) for i in ident]
  File "./dump-to-bucket.py", line 62, in <listcomp>
    ids = [join_ident_hash(*i) for i in ident]
TypeError: join_ident_hash() missing 1 required positional argument: 'version'
```

It needs to set `portion` to `-1` like the other `get_content` calls before this one.

File: dump/dump-to-bucket.py (hunk 4 above)
Comment (karenc) [reply dump/dump-to-bucket.py:62]: This conversation is moving to line 246

File: dump/dump-to-bucket.py (hunk 5 above)
Comment (m1yag1) [reply dump/dump-to-bucket.py]: What line? Could you push it so I don't have to decipher what your talking about?

Comment (m1yag1) [reply dump/dump-to-bucket.py]: n/m figured it out.

Comment (m1yag1) [reply dump/dump-to-bucket.py]: but then i hit another error

```
File "/Users/mike/Projects/00_openstax/00_content_engineering/rap-spike-lambda/dump/dump-to-bucket.py", line 270, in dump_in_bucket
    key = gen_filepath(type, ident)
  File "/Users/mike/Projects/00_openstax/00_content_engineering/rap-spike-lambda/dump/dump-to-bucket.py", line 78, in gen_filepath
    return func(ids)
  File "/Users/mike/Projects/00_openstax/00_content_engineering/rap-spike-lambda/dump/dump-to-bucket.py", line 73, in <lambda>
    'baked-page-json': lambda i: f'baked/{i[0]}:{i[1]}.json',
IndexError: list index out of range
```

**Comment by mmulich** [comment]
It looks like I committed the wrong version. So I'll need to back out some `get_content` function changes.

File: dump/dump-to-bucket.py (hunk 2 above)
Comment (mmulich) [reply dump/dump-to-bucket.py:206]: Yeah, I copied the wrong version here. I'll need to take out line 213 and uncomment the stuff above it. You can see this repeatedly in the next few statements too. I tried to refactor and forgot I needed access to the response.

File: dump/requirements.txt (hunk 3 above)
Comment (mmulich) [reply dump/requirements.txt:5]: O... my. The sloppiness of all this is terrible. 😝

**Review by m1yag1 — APPROVED** [review]

**Review by karenc — APPROVED** [review]
The script hasn't completed yet for me :joy: but I'm happy with it.

**Review by therealmarv — APPROVED** [review]
now works for me too.

**Pull request merged**